### PR TITLE
ISSUE-389 # fix retry report issue

### PR DESCRIPTION
**Pull request by sparrowV**
# fix retry issue

PR Branch
**https://github.com/sparrowV/zerocode**

## Checklist:

* [x] Unit tests added?

* [ ] Integration tests added?

* [x] Test names are meaningful?

* [x] Feature manually tested?

* [x] Branch build passed?

* [x] No 'package.*' in the imports

* [ ] Relevant Wiki page updated with clear instruction for the end user

* [ ] Http test added to `http-testing` module(if applicable) ?

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?

### CODE REVIEW

File: core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java [hunk 1]
```diff
@@ -82,6 +80,26 @@ public ZeroCodeReportGeneratorImpl(ObjectMapper mapper) {
         this.mapper = mapper;
     }
 
+    /**
+     * gets unique steps from a scenario. In case of retries, the steps have same correlation id and if
+     * one of the retries is successful, we include it in the result(not the one which failed).
+     * @param steps
+     * @return
+     */
+    private List<ZeroCodeReportStep> getUniqueSteps(List<ZeroCodeReportStep> steps){
```
Comment (authorjapps) [inline core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java:89]: Sounds correct. 👍 
Can you apply the same to the CSV results too.

The changes to be applied here:

> org.jsmart.zerocode.core.report.ZeroCodeReportGeneratorImpl#buildCsvRows

Comment (sparrowV) [reply core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java:90]: @authorjapps  sure I added for csv too

**Comment by authorjapps** [comment]
Solves #389

File: core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java [hunk 2]
```diff
@@ -82,6 +81,26 @@ public ZeroCodeReportGeneratorImpl(ObjectMapper mapper) {
...
```
Comment (authorjapps) [inline core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java:90]: Thanks! Looks great!

Can you add some unit tests for this `getUniqueSteps(...)` please?
That's covering some RETRY and non-RETRY cases?

File: core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java [hunk 3]
```diff
@@ -82,6 +81,26 @@ public ZeroCodeReportGeneratorImpl(ObjectMapper mapper) {
...
+     List<ZeroCodeReportStep> getUniqueSteps(List<ZeroCodeReportStep> steps){
+        Map<String,ZeroCodeReportStep> result = new HashMap<>();
+        steps.forEach(step->{
+            if(result.containsKey(step.getCorrelationId())){
+                if(step.getResult().equals(RESULT_PASS)){
+                    result.put(step.getCorrelationId(),step);
+                }
+            }else{
+                result.put(step.getCorrelationId(),step);
+            }
```
Comment (Cyril-Edme) [inline core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java:99]: I think that could be simplified with a merge, something like that :
`result.merge(step. getCorrelationId(), step, (s1, s2) -> RESULT_PASS.equals(s1.getResult()) ? s1 : s2)`

Comment (authorjapps) [reply core/src/main/java/org/jsmart/zerocode/core/report/ZeroCodeReportGeneratorImpl.java:99]: @sparrowV , this looks clean, short and not too complicated to read!
If you could take care fo this(`result.merge(...)`) bit , please. We are almost there ✌️

**Pull request merged**